src/nacl/bindings/crypto_secretstream.py

```diff
@@ -0,0 +1,271 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+from nacl import exceptions as exc
+from nacl._sodium import ffi, lib
+from nacl.exceptions import ensure
+
+
+crypto_secretstream_xchacha20poly1305_ABYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_abytes()
+crypto_secretstream_xchacha20poly1305_HEADERBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_headerbytes()
+crypto_secretstream_xchacha20poly1305_KEYBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_keybytes()
+crypto_secretstream_xchacha20poly1305_MESSAGEBYTES_MAX = \
+    lib.crypto_secretstream_xchacha20poly1305_messagebytes_max()
+crypto_secretstream_xchacha20poly1305_STATEBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_statebytes()
+
+
+crypto_secretstream_xchacha20poly1305_TAG_MESSAGE = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_message()
+crypto_secretstream_xchacha20poly1305_TAG_PUSH = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_push()
+crypto_secretstream_xchacha20poly1305_TAG_REKEY = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_rekey()
+crypto_secretstream_xchacha20poly1305_TAG_FINAL = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_final()
+
+
+def crypto_secretstream_xchacha20poly1305_keygen():
+    """
+    Generate a key for use with
+    :func:`.crypto_secretstream_xchacha20poly1305_init_push`.
+
+    """
+    keybuf = ffi.new(
+        "unsigned char[]",
+        crypto_secretstream_xchacha20poly1305_KEYBYTES,
+    )
+    lib.crypto_secretstream_xchacha20poly1305_keygen(keybuf)
+    return ffi.buffer(keybuf)[:]
+
+
+def crypto_secretstream_xchacha20poly1305_init_push(key):
+    """
+    Initialize a crypto_secretstream_xchacha20poly1305 encryption buffer.
+
+    :param key: must be
+                :data:`.crypto_secretstream_xchacha20poly1305_KEYBYTES` long
+    :type key: bytes
+    :return: (statebuf, header)
+    :rtype: (bytes, bytes)
+    """
+    ensure(
+        isinstance(key, bytes),
+        'Key must be a bytes sequence',
+        raising=exc.TypeError,
+    )
+    ensure(
+        len(key) == crypto_secretstream_xchacha20poly1305_KEYBYTES,
+        'Invalid key length',
+        raising=exc.ValueError,
+    )
+
+    statebuf = ffi.new(
+        "unsigned char[]",
+        crypto_secretstream_xchacha20poly1305_STATEBYTES,
+    )
+
+    # both _salt and _personal must be zero-padded to the correct length
+    headerbuf = ffi.new(
+        "unsigned char []",
+        crypto_secretstream_xchacha20poly1305_HEADERBYTES,
+    )
+
+    rc = lib.crypto_secretstream_xchacha20poly1305_init_push(
+        statebuf, headerbuf, key)
+    ensure(rc == 0, 'Unexpected failure', raising=exc.RuntimeError)
+
+    return statebuf, ffi.buffer(headerbuf)[:]
+
+
+def crypto_secretstream_xchacha20poly1305_push(statebuf, m, ad, tag):
```

Comment:
I was looking through some of our other bindings, and it looks to me like for things like ``ad`` and ``tag`` that are "optional" (to the extent anything is optional in C) or that have a sane default value, we've been setting default values for them. So I think changing this to having ``statebuf, m, ad=None, tag=0`` would be a good improvement.

Reply:
Part of me wonders if turning ``tag`` into an integer based Enum wouldn't be a horrible idea. I can't find any other APIs in our bindings that have a similar flag like this to crib off of. I suppose leaving it as an integer that you can pass one of the defined constants into is the truest Python to C conversion that these bindings typically try to be.

Reply:
I think it's ok to change the defaults as you suggested but with respect to the enum - I think leaving the bindings coupled pretty tightly with the libsodium documentation is handy since pynacl itself hasn't been attempting to expose any docs for the binding layer so far.

Perhaps an enum might be exposed in a higher level stream api, if one is defined in the future?

Reply:
Yea, that seems reasonable to me.